[또링] 로또 2차 미션 코드리뷰 요청합니다.

jnsorn opened the pull request:
2차 미션 리뷰도 잘 부탁드립니다! 😊

leewnsdud committed:
docs : README.md 작성

leewnsdud committed:
feat : ConsoleUILottoApplication 클래스 추가

leewnsdud committed:
feat : Money객체 추가 및 테스트코드 작성

 1. 정수만 입력 가능
 2. 음수와 0은 입력 불가능
 3. 1,000원 단위로만 입력 가능
 4. 최대 구입 가능 금액 100,000원 제한

leewnsdud committed:
fest : Number객체 추가 및 테스트코드 작성

leewnsdud committed:
refactor : Number객체를 캐싱하는 방식으로 변경

 1. 생성자 대신 valueOf로 변경
 2. 테스트코드의 생성자를 valueOf로 수정

leewnsdud committed:
feat : Lotto 객체 추가 및 테스트코드 작성

leewnsdud committed:
feat : 구입 금액을 입력받는 기능 추가

leewnsdud committed:
refactor : ConsoleUIApplication에서 LottoController를 호출하는 방식으로 변경

leewnsdud committed:
feat : 구매 가능한 로또의 개수를 구하는 기능 및 테스트코드 추가

leewnsdud committed:
refactor : ConsoleUIApplication의 receiveInputMoney가 Money 객체를 반환하도록 변경

leewnsdud committed:
feat : 구매한 로또의 개수를 출력하는 기능 구현

leewnsdud committed:
refactor : Exception 클래스를 exception패키지로 이동

leewnsdud committed:
refactor : Lotto의 사이즈를 검증하는 기능 추가 및 테스트코드 변경

 - 기존의 Empty를 검증하는 기능 제거

leewnsdud committed:
feat : 6개의 랜덤 숫자로 로또를 생성하는 기능 구현

leewnsdud committed:
feat : 구매한 로또를 출력하는 기능 구현

leewnsdud committed:
refactor : LottoController의 start 메소드에서 List<Lotto>를 반환하도록 변경

leewnsdud committed:
refactor : Money 객체와 Number 객체의 이름 변경

leewnsdud committed:
refactor : view클래스 이름을 구체적으로 변경

leewnsdud committed:
refactor : 로또를 생성하는 방식을 캐싱 된 리스트를 불러와 셔플하는 방식으로 변경

leewnsdud committed:
refactor : exception의 패키지 이동 및 이름 변경

leewnsdud committed:
feat : 6개의 당첨 번호를 입력하는 기능 구현

leewnsdud committed:
feat : 입력한 6개의 당첨 번호로 LottoNumber객체 리스트들을 생성하는 기능 추가

leewnsdud committed:
feat : 1개의 보너스 번호를 입력하는 기능 구현

leewnsdud committed:
feat : 당첨 번호와 중복된 보너스 번호를 검증하는 기능 추가

leewnsdud committed:
refactor : WinningLotto 객체 생성 및 예외처리 부분을 메서드 분리

leewnsdud committed:
feat : 구입한 로또와 당첨 번호의 일치 여부를 찾아내는 기능 구현

leewnsdud committed:
feat : 당첨통계 안내 메세지를 출력하는 기능 구현

leewnsdud committed:
refactor : LottoRank 객체의 메소드 분리 및 테스트 추가

leewnsdud committed:
refactor : 전체적인 메소드와 상수 이름 변경

leewnsdud committed:
feat : 당첨 통계를 출력하는 기능 구현

leewnsdud committed:
refactor : 당첨 통계를 0개를 포함한 오름차순 출력으로 변경

leewnsdud committed:
feat : LottoMoney를 연산하는 기능 및 테스트코드 구현

leewnsdud committed:
feat : 총 수익률을 출력하는 기능 구현

leewnsdud committed:
refactor : 사용하지 않는 클래스 제거

jnsorn committed:
Merge pull request #1 from lxxjn0/lxxjn0

jnsorn committed:
refactor : 사용하지 않는 클래스 정리

jnsorn committed:
refactor : 요구조건에 맞게 ConsoleOutputView 변경

jnsorn committed:
refactor : 전반적인 접근 지정자 변경 및 패키지 분리

jnsorn committed:
style : 4스페이스에서 Tab으로 변경

jnsorn committed:
refactor : Rank와 관련된 상수를 LottoRank로 이동

jnsorn committed:
refactor : 로직이 들어간 메서드들의 이름을 get에서 calculate로 변경

jnsorn committed:
refactor : view 클래스의 기본 생성자를 private으로 변경

jnsorn committed:
refactor : 요구사항에 명시되지 않은 내용 삭제 및 가독성을 위한 식별자 추가

jnsorn committed:
refactor : LottoNumber 자료형을 list에서 set으로 변경

jnsorn committed:
refactor : Lotto list를 Lottos라는 일급컬렉션으로 변경

jnsorn committed:
refactor : test 패키지 분리

jnsorn committed:
refactor : package 이름 변경

jnsorn committed:
refactor : WinningLotto 클래스를 생성하여 당첨 번호와 보너스 번호를 한 곳에서 관리

jnsorn committed:
refactor : result 패키지를 생성하여 패키지를 분리

jnsorn committed:
refactor : LottoRankRecord 클래스를 생성하여 로또들의 순위를 기록

jnsorn committed:
refactor : LottoNumber 클래스 내에 있는 캐시를 분리하여 LottoNumberCache생성

jnsorn committed:
style : 모든 코드에 네이버 컨벤션 적용

jnsorn committed:
refactor : ConsoleUILottoApplication 코드를 controller로 옮김

jnsorn committed:
refactor : controller에 있던 수익률 계산 함수를 LottoWinningResult로 이동

jnsorn committed:
refactor : controller의 lottos를 생성하는 기능을 LottoGenerator으로 이동

jnsorn committed:
refactor : controller에서 winningLotto 받는 기능을 메서드로 분리

jnsorn committed:
refactor : 로또 번호 캐시 상수명 변경

jnsorn committed:
fix : 로또 번호 5개 일치, 보너스 번호 불일치시 3등이 당첨되도록 수정

jnsorn committed:
refactor : 모호한 메서드 명 변경 및 중복 상수 제거

jnsorn committed:
refactor :  메서드 역할 분리 및 리팩토링

calculate 함수가 계산만 하도록 초기화 작업 기능은 생성자로 분리

jnsorn committed:
refactor : 메서드 역할 분리 및 리팩토링

jnsorn committed:
Merge remote-tracking branch 'origin/jnsorn' into jnsorn

jnsorn committed:
docs : 2단계 구현 기능 목록 작성

jnsorn committed:
docs : 구현 기능 목록 추가

jnsorn committed:
feat : 수동으로 구매할 로또 수를 입력 받는 기능

jnsorn committed:
refactor : 로또 숫자와 개수를 구분하기 위해 변수 및 함수의 이름 변경

jnsorn committed:
feat : 수동으로 구매할 번호를 입력 받는 기능

jnsorn committed:
feat : 생성된 로또를 출력하는 기능

jnsorn committed:
fix : 수동로또와 자동로또를 통합한 결과를 내도록 수정

kingbbode commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -13,33 +10,69 @@
+        if (numberOfManualLotto < 0 || numberOfManualLotto > numberOfLotto) {
+            throw new IllegalArgumentException(COUNT_OF_MANUAL_IS_WRONG_RANGE_MESSAGE);
+        }
+        return numberOfManualLotto;
```
입력에 대한 검증과 메시지는 입력에서 처리되어야 하지 않을까요?

jnsorn replied:
네 해당 부분 수정하겠습니다! :)

kingbbode commented on src/main/java/lotto/domain/lotto/Lottos.java:
```diff
@@ -3,6 +3,7 @@
 import java.util.List;
+import java.util.Objects;
 
 public class Lottos implements Iterable<Lotto> {
```
일급컬렉션으로서 하는 역할이 없네요 :)
그렇다면 List<Lotto> 를 사용하는게 더 편하지 않을까요?
의미없는 객체 포장은 불필요해보여요.

jnsorn replied:
넵 해당 부분 수정하였습니다! 확실히 List로 사용하는 것이 더 편하게 느껴지네요 😀

kingbbode commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -13,33 +10,69 @@
+        WinningLotto winningLotto = receiveWinningLotto();
+        LottoWinningResult winningResult = new LottoWinningResult(totalLottos, winningLotto);
+        printWinningResult(winningResult.getLottoRankCount());
+        printWinningRatio(winningResult.calculateWinningRatio(inputLottoMoney));
```
`LottoMoney`, `Lottos`, `Lotto`, `WinningLotto` 등 여러 객체를 추출했지만,
객체 간의 관계는 설계되지 못하고 `LottoController` 에서 나열로만 처리되고 있네요 :) 
`LottoController` 가 너무 많은 책임과 역할을 가지고 있다고 생각이되요.

로또의 구입을 담당하는 `LottoMachine` 을 추출하여, `LottoMachine` 을 통해 로또를 생성해낼 수 있을거에요.
`LottoMoney` 는 현재 금액을 내장하고 있지만, 로또의 가격까지도 알고 있기 때문에 책임이 모호한 객체가 되어있어요.
`Money` 는 돈을 관리하는 역할에 집중하고, 로또의 가격을 관리하는 어딘가에서 금액을 차감해가며, 구입할 수 있는 개수를 책임질 수 있을 것 같아요. 

위에는 예시로 역할과 책임을 더 작은 단위로 나누고 객체 간의 관계를 만들 수 있지 않을까하여 적어보았어요.

`LottoController` 가 가지고 있는 너무 많은 역할을 덜어줄 수 있도록 리펙토링해보는 것은 어떨까요?

jnsorn replied:
좋은 리뷰 감사합니다 :) 사실 어떻게 리팩토링을 해야할까 고민하다가 마감시간이 되어 우선 올려야겠다는 마음에 제출했는데 이렇게 빨리 리뷰를 해주실 줄 몰랐습니다... (정말 돌아가는 쓰레기 코드..같아서 매우 부끄럽습니다...🤣) 우선 리뷰어님이 말씀해주신 내용들 바탕으로 열심히 리팩토링 해보고 다시 리뷰요청 보내겠습니다!

kingbbode reviewed (changes requested):
안녕하세요 :)
코멘트 작성해두었습니다.
확인 부탁드려요!

jnsorn committed:
refactor : 각 책임을 가지는 객체를 생성하여 컨트롤러의 역할을 최소화

LottoMachine, AutoLottoGenerator, ManualLottoGenerator, CountOfManualLotto, LottoPrice, NumberLinesOfManualLotto 객체 생성

jnsorn committed:
refator : Lottos라는 이름을 가진 변수와 함수의 이름을 LottoTicket으로 변경

jnsorn committed:
refactor : 티켓을 구매할 때 전달하는 인자를 변경

- 기존 수동 로또 개수와 수동 로또 번호들을 전달하는 방법에서 수동 로또 번호들만 전달하고 사이즈를 얻어오는 방법으로 변경

jnsorn committed:
refactor : OutputView 메서드의 변수명 수정

jnsorn committed:
refactor : LottoMachine을 정적으로 생성하도록 변경

jnsorn committed:
fix : LottoMachine 테스트 수정

jnsorn committed:
test : LottoPrice 객체 테스트 추가

jnsorn committed:
test : CountOfManualLottoTicket 테스트 추가

jnsorn committed:
refactor : NumberLinesOfManualLotto 클래스의 메서드 순서 변경 및 테스트 추가

jnsorn committed:
test : ManualLottoGenerator 테스트 추가

jnsorn committed:
test : AutoLottoGenerator 테스트 추가

jnsorn committed:
refactor : WinningLotto 생성 방식 변경 

기존 LottoParser로 생성하던 방법에서 ManualLottoGenerator로 생성하는 방법으로 변경

jnsorn committed:
refactor : 필요없는 일급컬렉션(LottoTicket) 제거

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,41 @@
+import lotto.domain.lotto.Lotto;
+import lotto.domain.number.NumberLinesOfManualLotto;
+
+public class LottoMachine {
```
정적 클래스로 제공하는 이유는 무엇일까요?
더 유연하게 사용할 수 있도록 객체로서 제공되는 것이 더 효율적이지 않을까요?

jnsorn replied:
LottoMachine 클래스는 단순히 금액을 넣으면 로또를 반환하는 역할을 한다고 생각하여 정적 클래스로 만들었습니다😥 하지만 컨트롤러의 책임을 분리하면서 LottoMachine의 생성자를 통해 금액과 수동 로또 갯수를 입력받도록 하여 LottoMachine 객체를 더 유연하게 사용할 수 있도록 해당 부분 수정하였습니다😀

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,41 @@
+		countOfAutoLottoTicket = DEFAULT_COUNT;
+	}
+
+	public static List<Lotto> buyLottoTicket(int countOfAllLotto, NumberLinesOfManualLotto manualLottoNumbers) {
```
`LottoMachine` 의 역할은 무엇인가요?
정보를 소유하고 가지고 있다면 개방되지 않아도 될 데이터들이 많이 있을 것 같아요 :)
단순하게 비지니스 로직만을 책임 지기 때문에 `LottoController` 에서 더 많은 정보를 요구하고 책임져야할 기능들이 많아진 것 같아요.

적절한 응집으로 `LottoController` 가 가지고 있는 책임과 역할을 분리해보는 것은 어떨까요?

kingbbode commented on src/main/java/lotto/domain/lotto/CountOfManualLottoTicket.java:
```diff
@@ -0,0 +1,41 @@
+package lotto.domain.lotto;
+
+public class CountOfManualLottoTicket {
```
이 객체의 역할과 책임은 무엇인가요?
행위를 포장해놓은 객체는 의미가 없지 않을까요?
대신 이 행위를 가지고 있어야할 객체를 찾아보는 것도 좋을 것 같습니다 :)

kingbbode commented on src/main/java/lotto/domain/money/LottoPrice.java:
```diff
@@ -0,0 +1,9 @@
+package lotto.domain.money;
+
+public class LottoPrice {
```
정적 클래스라면 `private 생성자` 로 생성을 막아주는 것이 안전한 방법입니다 :)

kingbbode replied:
이 클래스도 행위를 책임질 객체를 찾지 못하였거나 추출하지 못했기 때문에 애매하게 만들어진 역할을 가지고 있네요 :)

kingbbode commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -3,43 +3,53 @@
 import lotto.domain.result.LottoWinningResult;
 import lotto.domain.result.WinningLotto;
 
 public class LottoController {
```
각 클래스의 역할과 책임을 분리해보면 좋을 것 같습니다.
책임과 역할을 나누었다면 그것을 클래스로 작성하여 `LottoController` 에서는 큰 행위만 수행하도록 해보시고,
두번째로 리펙토링을 하며, 나누어진 클래스 안에서 책임과 역할을 또 다시 분리할 수 있는지를 보시길 바랍니다.

kingbbode reviewed (changes requested):
안녕하세요 :)
아쉬운 점이 있어 코멘트 작성하였습니다.
확인 부탁드려요!

jnsorn committed:
refactor :  각 객체가 적절한 역할과 책임을 가지게 하고, controller의 역할을 최소화

jnsorn committed:
test : 테스트 메서드 이름 변경 및 에러메세지를 통한 에러확인 구체화

jnsorn committed:
refactor : 수동 로또 생성 방식을 변경하여 controller에서 winningLotto생성 책임을 분리

jnsorn commented:
저번 리팩토링을 하며 새로 만든 ``LottoMachine``을 통해 나름 controller가 하는 역할을 덜어주었다고 생각했는데, 다시 리뷰를 받고 곰곰히 생각해보니 제 착각이었던 것 같습니다😅 사실 저번주와 똑같은 리뷰를 받고 나서 좀 막막해서 다시 리뷰요청 보내기까지 시간이 좀 걸린 것 같아요...😥 

나름 controller의 역할을 덜어준다고 코드를 다시 짰는데 여기서 더 어떻게 할 수 있을지에 대해서 계속해서 생각을 해봤더니, 정말 말 그대로 **controller는 입력받은 결과를 ``LottoMachine``에 넘겨 로또에 관한 책임을 지도록 하는 것이 맞겠다**는 생각을 했습니다. 예를 들어 로또 갯수를 계산하거나, 로또티켓을 생성하거나 하는 그런 행위들이요! 
말씀해주신 것처럼 ``LottoMachine``내에서 돈을 검증해서 생성하는 역할, 로또를 검증해서 생성하는 역할은 다시 ``Money``나 ``LottoGenerator``로 분리시키면 되겠다고 생각했습니다. 

이렇게 생각을 하고 나니 지적해주셨던 ``LottoPrice``는 정말 불필요한 객체라고 생각이 들었습니다. 

아직도 갈 길이 먼 것 같지만..!!  부족한 부분 많이 지적해주시면 더 열심히 수정해보겠습니다~😊

kingbbode commented on src/main/java/lotto/domain/money/Money.java:
```diff
@@ -45,15 +47,11 @@ private void validatePositive(long parsedMoney) {
 	}
 
 	private void validateUnit(long parsedMoney) {
```
Money 객체는 얼마를 몇개 구매할 수 있는지를 제공해주어 로또와의 의존을 분리했다면 더 투명한 객체가 되었을 것 같네요 :)

kingbbode commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -3,25 +3,41 @@
+		countOfAutoLottoTicket = countOfAllLotto - countOfManualLottoTicket;
 
 		ManualLottoTicketGenerator manualLottoTicketGenerator = new ManualLottoTicketGenerator(manualLottoNumbers);
 		AutoLottoTicketGenerator autoLottoTicketGenerator = new AutoLottoTicketGenerator(countOfAutoLottoTicket);
```
ManualLottoTicketGenerator 와 AutoLottoTicketGenerator 가 위 처럼 되어있다면 매번 생성하는 비효율을 갖고,
인터페이스를 갖는 이점을 모두 갖을 수 없겠네요 :)

인터페이스를 추출한다는 것은 사용하는 쪽에서 구현을 모르게 하여 결합을 분리하겠다는 설계 의도를 갖습니다.

kingbbode reviewed (approved):
코드가 많이 깔끔해졌네요 :) 고생하셨어요! 👍 
결합을 낮추는 연습을 앞으로도 많이 하시면 좋겠네요!
이만 머지하겠습니다!
(코멘트도 확인 부탁드려요)
